### PR TITLE
Adds feature to declare optional services via @?service_name

### PR DESCRIPTION
**Pull request by fruit**
Allows to declare services as non-required dependency in `services.yml` file.

This feature is also implemented in [Symfony 2](http://symfony.com/doc/current/book/service_container.html#ignoring-missing-dependencies).

Example:

``` yaml
services:
  opt_args:
    class: ClassOptionalArguments
    arguments: ['@?bar']
```

The `@bar` service is missing in service container, but the usage of `opt_args` service will not throw the `InvalidArgumentException`.

``` php
$optArgsService = $container->getService('opt_args');
// exception is not thrown
```

### CODE REVIEW

File: lib/service/sfServiceContainerDumperPhp.class.php [hunk 1]
```diff
@@ -351,6 +368,10 @@ protected function getServiceCall($id)
       return '$this';
     }
 
+    if (0 === strpos($id, '?')) {
```
Comment (j0k3r) [inline lib/service/sfServiceContainerDumperPhp.class.php]: Could you move the `{` to the new line?

File: lib/service/sfServiceContainerDumperPhp.class.php [hunk 2]
```diff
@@ -143,7 +143,24 @@ protected function addServiceConfigurator($id, $definition)
...
+      \$configurator->{$callable[1]}(\$instance);
+    }
+
+BLOCK
+;
+        }
+        else
+        {
+          return sprintf("    %s->%s(\$instance);\n", $this->getServiceCall($refId), $callable[1]);
+        }
```
Comment (j0k3r) [inline lib/service/sfServiceContainerDumperPhp.class.php]: This `else` isn't necessary, you can remove it and directly use the `return`.

**Comment by fruit** [comment]
I've squeezed fixes into one single commit.

**Comment by j0k3r** [comment]
Thanks! This looks very interesting.
Maybe @GromNaN can have a look too?

**Comment by GromNaN** [comment]
I've never used such feature. 
The implementation looks correct.

**Comment by fruit** [comment]
At the moment, the only known problem is that `sfServiceContainerDumperGraphviz` renders dependencies with a leading "?" as missing.

**Comment by j0k3r** [comment]
@fruit do you plan to fix it then?

**Comment by fruit** [comment]
@j0k3r I've fixed it, but did it without tests at the moment.

**Comment by fruit** [comment]
All affected and new tests passes.

```bash
php test/unit/task/service/sfServicesGraphvizDumpTaskTest.php
find test/unit/service/ -name "*Test.php" -exec php {} \;
```

**Comment by fruit** [comment]
Perhaps, it is worth adding a documentation about this feature and new task...

**Comment by fruit** [comment]
@j0k3r @GromNaN any comments/questions/suggestions?

**Comment by GromNaN** [comment]
[👍](https://github.com/LExpress/symfony1/pull/126#issuecomment-222483508)

**Pull request closed without merging**